libgadget/density.c

```diff
@@ -449,6 +446,8 @@ density_reduce(int place, TreeWalkResultDensity * remote, enum TreeWalkReduceMod
     else if(P[place].Type == 5)
     {
         TREEWALK_REDUCE(BHP(place).Density, remote->Rho);
+        int bhpi = P[place].PI + SlotsManager->info[0].size * DENSITY_GET_PRIV(tw)->DoEgyDensity;
```

Comment:
I feel I have seen this type of access to 'temporal/transient' variables several times. What about introducing a TemporalSlots class to capture these tw->??? variables?

```
struct TemporalSlots {
union {
  double *,
  int *,
};
int NPerType[6];
};

tw->DhsmlDensityFactor = temporal_slots_new_double(int NPerType[6], ...); 
```

and (perhaps a macro ...)
```
double * pvalue = temporal_slots_get_double(tw->DhsmlDensityFactor, type, P[place].PI);
```

The advantage is that the logic for relative indexing no longer need to explicitly do a branching per access (and thus no need to maintain the consistency).

Reply:
We are certainly growing new boilerplate around this so something would be good. Could you expand on what your suggestion is?

I don't think branching is really a problem (these branches should be easily predicted), although a fairly expensive thing in the current code is the non-linear memory access inherent in the slots having a different ordering than the main particle table.